source/containerimage/pull.go

```diff
@@ -100,136 +100,141 @@ func (is *imageSource) Resolve(ctx context.Context, id source.Identifier, sm *se
+	ctx, done, err := leaseutil.WithLease(ctx, p.LeaseManager, leaseutil.MakeTemporary)
 	if err != nil {
-		return "", false, err
+		return "", nil, false, err
 	}
+	defer done(ctx)
+
+	resolveProgressDone := oneOffProgress(ctx, "resolve "+p.Src.String())
+	defer func() {
+		resolveProgressDone(err)
+	}()
+
+	metadata, err := p.PullMetadata(ctx)
+	if err != nil {
+		return "", nil, false, err
+	}
+
+	optSet := cache.AsOptSet(metadata.Remote, progress.CallbacksOf(ctx), metadata.Ref)
+
+	desc := metadata.MainManifestDesc
 	if index == 0 || desc.Digest == "" {
 		k, err := mainManifestKey(ctx, desc, p.Platform)
 		if err != nil {
-			return "", false, err
+			return "", nil, false, err
 		}
-		return k.String(), false, nil
+		return k.String(), optSet, false, nil
 	}
 	ref, err := reference.ParseNormalizedNamed(p.Src.String())
 	if err != nil {
-		return "", false, err
+		return "", nil, false, err
 	}
 	ref, err = reference.WithDigest(ref, desc.Digest)
 	if err != nil {
-		return "", false, nil
+		return "", nil, false, err
 	}
-	_, dt, err := imageutil.Config(ctx, ref.String(), p.Resolver, p.ContentStore, p.LeaseManager, &p.Platform)
+	_, dt, err := imageutil.Config(ctx, ref.String(), p.Resolver, p.ContentStore, p.LeaseManager, p.Platform)
 	if err != nil {
-		return "", false, err
+		return "", nil, false, err
 	}
 
 	k := cacheKeyFromConfig(dt).String()
 	if k == "" {
 		k, err := mainManifestKey(ctx, desc, p.Platform)
 		if err != nil {
-			return "", false, err
+			return "", nil, false, err
 		}
-		return k.String(), true, nil
+		return k.String(), optSet, true, nil
 	}
-	return k, true, nil
+
+	return k, optSet, true, nil
 }
 
 func (p *puller) Snapshot(ctx context.Context) (ir cache.ImmutableRef, err error) {
-	layerNeedsTypeWindows := false
-	if platform := p.Puller.Platform; platform != nil {
-		if platform.OS == "windows" && runtime.GOOS != "windows" {
-			ctx = winlayers.UseWindowsLayerMode(ctx)
-			layerNeedsTypeWindows = true
-		}
-	}
-
-	// workaround for gcr, authentication not supported on blob endpoints
-	pull.EnsureManifestRequested(ctx, p.Puller.Resolver, p.Puller.Src.String())
-
 	ctx, done, err := leaseutil.WithLease(ctx, p.LeaseManager, leaseutil.MakeTemporary)
 	if err != nil {
 		return nil, err
 	}
 	defer done(ctx)
 
-	pulled, err := p.Puller.Pull(ctx)
+	metadata, err := p.PullMetadata(ctx)
 	if err != nil {
 		return nil, err
 	}
-	if len(pulled.Layers) == 0 {
+
+	if len(metadata.Remote.Descriptors) == 0 {
 		return nil, nil
 	}
 
-	extractDone := oneOffProgress(ctx, "unpacking "+pulled.Ref)
 	var current cache.ImmutableRef
 	defer func() {
 		if err != nil && current != nil {
 			current.Release(context.TODO())
 		}
-		extractDone(err)
 	}()
-	for _, l := range pulled.Layers {
-		ref, err := p.CacheAccessor.GetByBlob(ctx, l, current, cache.WithDescription("pulled from "+pulled.Ref))
-		if err != nil {
-			return nil, err
+
+	var parent cache.ImmutableRef
+	for _, layerDesc := range metadata.Remote.Descriptors {
```

Comment:
afaics the actual remote is ignored and the one coming through side-channel via context is used instead. We should only use side-channel for cache load where it is needed and not leak the context magic to `cache` package.